### ISSUE TITLE
async logger in realtime applications

### ISSUE THREAD

**Issue by briansoe66** [gabime/spdlog#1873]
Awesome library, thanks so much! I am interested in using asynchronous logger with multi sink in a realtime application.
1. If async_logger creates a new thread, does the stdout sink and rotating file sink still need to be multithreaded to be used with the asynchronous logger, like shown in the README?
2. Can i use the init_thread_pool() on_thread_start function to set the iostream / fstream thread to be lower priority than my realtime thread? for example:
```
void on_thread_start{  
  struct sched_param param;
  param.sched_priority = 40; // lower priority
  sched_setscheduler(0, policy, &param);
}

int main(){
struct sched_param param;
param.sched_priority = 99; // high priority
sched_setscheduler(0, policy, &param);

spdlog::init_thread_pool(8192, 1, on_thread_start);
// do the rest of the log initialization from the "asynchronous logger with multi sink" example in the README

// run some realtime code, that does non-blocking logging
}
```
3. Does the async implementation of spdlog prevent priority inversion?
4. If not, is it on the roadmap?
5. If not, what class/mutex would I need to modify with PTHREAD_PRIO_INHERIT?
6. Should I create a PR, or that too niche or platform dependent?

**Comment by tt4g** [comment]
> 1. If async_logger creates a new thread, does the stdout sink and rotating file sink still need to be multithreaded to be used with the asynchronous logger, like shown in the README?

If the thread pool has only one worker, there is no need to use multithreaded sinks.
However, if you call `logger->flush()` to manipulate sinks from outside, use multi-threaded sinks to avoid state conflicts.
This is because some sinks protect flush operations with mutex.

https://github.com/gabime/spdlog/blob/de0dbfa3596a18cd70a4619b6a9766847a941276/include/spdlog/sinks/stdout_sinks-inl.h#L64-L69

**Comment by tt4g** [comment]
> 3. Does the async implementation of spdlog prevent priority inversion?

What does priority inversion mean?
If you mean the order of the log messages, the order is guaranteed as long as there is one worker thread in the thread pool.
Look wiki; https://github.com/gabime/spdlog/wiki/6.-Asynchronous-logging#message-ordering

**Comment by briansoe66** [comment]
`What does priority inversion mean?`
If the async logger queue implementation uses a mutex, then this can lead to [priority inversion ](https://en.wikipedia.org/wiki/Priority_inversion). 

Example implementation demonstrating priority inversion:

1. high-priority thread creates an async logger
2. async logger creates a worker thread
3. worker thread is set as lower priority
4. high-priority thread logs message 1. This locks the mutex, adds message 1 to the queue, and unlocks the mutex.
5. low-priority thread locks the mutex, dequeues message 1, and makes a copy. It is about to unlock the mutex when:
6. high-priority thread logs message 2. It tries to acquire the mutex, but cannot. The thread is blocked.
7. a medium-priority thread interrupts the low priority thread, to do a super long computations.
8. now both the high and low priority threads are blocked by this super long computations. This is priority inversion.
9. medium-priority thread finishes computation
10. low priority thread unlocks the mutex
11. high priority thread acquires the mutex, adds message 2 to queue, and unlocks the mutex. logging message 2 is complete.

**Comment by tt4g** [comment]
Hmmm. I don't think spdlog's asynchronous logger prevents priority inversion of threaded tasks.
There doesn't seem to be any code explicitly for that purpose.
As you presented, I think it would happen depending on the order in which the critical sections of the mutex are retrieved.

**Comment by gabime** [comment]
> if you call logger->flush() to manipulate sinks from outside, use multi-threaded sinks to avoid state conflicts.

If the logger is async, It is not required since ```logger->flush()``` just adds a "flush" request into the queue to be processed by the worker thread.

**Comment by briansoe66** [comment]
@gabime, i don't see any mutex lock when messages are pushed to the queue. Is the queue assumed to be long enough, that no mutex is needed? So the worker thread will never block the main thread?

**Comment by gabime** [comment]
There is mutex lock. see https://github.com/gabime/spdlog/blob/v1.x/include/spdlog/details/mpmc_blocking_q.h

**Comment by briansoe66** [comment]
Thanks, that should get me started. In mpmc_blocking_queue constructor, i'll try overriding the pthread_mutex_t returned by queue_mutex_.native_handle() with a PTHREAD_PRIO_INHERIT initialized pthread_mutex_t.

**Comment by gabime** [comment]
ok, enjoy.

**Comment by lbckmnn** [comment]
@briansoe66 did you manage to somehow get priority inheritance?
